### PR TITLE
virt-controller: always store evictionStrategy in Spec

### PR DESCRIPTION
**Pull request by acardace**
Store the VM assigned eviction strategy into the VMISpec, this
comes in handy as it's not always clear which evictionStrategy is
assigned to a VM if the cluster-wide setting was different
when the VM was first started.

https://bugzilla.redhat.com/show_bug.cgi?id=2228036

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by aburdenthehand** [comment]
cc @alaypatel07

**Review by vasiliy-ul — COMMENTED** [review]
This PR now includes the commit [instancetype: Refactor inferFromVolume functional tests](https://github.com/kubevirt/kubevirt/commit/7dd2fad5378273d45e180fe6531baeb40747348d) which is not referenced in the description and does not seem related to eviction strategy. Just wondering: is it included intentionally or by mistake?

UPD: I am fine with having the commit here, but then maybe it would be better to update the PR description.

**Comment by lyarwood** [comment]
> This PR now includes the commit [instancetype: Refactor inferFromVolume functional tests](https://github.com/kubevirt/kubevirt/commit/7dd2fad5378273d45e180fe6531baeb40747348d) which is not referenced in the description and does not seem related to eviction strategy. Just wondering: is it included intentionally or by mistake?
> 
> UPD: I am fine with having the commit here, but then maybe it would be better to update the PR description.

Yeah that's just a bad rebase, that landed a while ago in https://github.com/kubevirt/kubevirt/pull/9802 it doesn't need to be here

**Comment by vasiliy-ul** [comment]
Noted, thanks. The part with the eviction strategy looks good to me. It makes it easier to see what is currently used for the VM.

**Comment by acardace** [comment]
@vasiliy-ul ouch sorry, indeed a bad rebase! Thanks for catching it.

File: staging/src/kubevirt.io/api/core/v1/types.go [hunk 1]
```diff
@@ -280,6 +280,9 @@ type VirtualMachineInstanceStatus struct {
 	// Current topology may differ from the desired topology in the spec while CPU hotplug
 	// takes place.
 	CurrentCPUTopology *CPUTopology `json:"currentCPUTopology,omitempty"`
+
+	// EvictionStrategy shows the VM evictionStrategy
```
Comment (alaypatel07) [inline staging/src/kubevirt.io/api/core/v1/types.go:284]: is there a 
```
//+optional
```
required here? I see that its present for almost all fields that are optional except `CurrentCPUTopology`?

File: pkg/virt-controller/watch/vmi.go [hunk 2]
```diff
@@ -603,6 +604,9 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
 					}
 				}
 			}
+
+			vmiCopy.Status.EvictionStrategy = migrations.VMIEvictionStrategy(c.clusterConfig, vmi)
```
Comment (alaypatel07) [inline pkg/virt-controller/watch/vmi.go:608]: Instead of introducing a new status field, would it be better to default the EvictionStrategy to `kubevirtCR.spec.evictionStrategy` if `vmi.spec.evictionStrategy` is empty during vmi creation?

vmiCreation happens once, this will make sure the correct strategy is be persisted at creation time. Here is an example of another field which does the same: https://github.com/kubevirt/kubevirt/blob/25d2e3329577d90a764d127b947e2608de28871f/pkg/virt-api/webhooks/defaults.go#L215, introduced in this PR: https://github.com/kubevirt/kubevirt/pull/8293

Comment (vasiliy-ul) [reply pkg/virt-controller/watch/vmi.go:608]: I guess this will change the behavior: now if a `vmi.spec.evictionStrategy` is not set, the **current** value from KubeVirt CR is used. If the VMI value is defaulted, then it will be fixed as if the user set it on creation time.

Comment (alaypatel07) [reply pkg/virt-controller/watch/vmi.go:608]: @vasiliy-ul the defaulting will only occur  if vmi.spec.evictionStrategy is not set. It's the same behavior. Thats what happens with `vmi.spec.architecture` as well. So I dont think it will be a behavior change

File: staging/src/kubevirt.io/api/core/v1/types.go (hunk 1 above)
Comment (acardace) [reply staging/src/kubevirt.io/api/core/v1/types.go:284]: The field always gets set so I'd say no? Do you think is required?

File: pkg/virt-controller/watch/vmi.go (hunk 2 above)
Comment (acardace) [reply pkg/virt-controller/watch/vmi.go:608]: @alaypatel07 yes I think this is a great idea! Thank you for the suggestion, I will update the PR.

Comment (vasiliy-ul) [reply pkg/virt-controller/watch/vmi.go:608]: > So I dont think it will be a behavior change

Still, I am not sure. Looking at the code:

https://github.com/kubevirt/kubevirt/blob/0a91f46054c6429cd4311e4ecc255bdefe442b2d/pkg/util/migrations/migrations.go#L65-L71


So, the current flow:
- let's say, initially the eviction strategy in KubeVirt CR is `EvictionStrategyNone`
- I create a VM with an empty eviction strategy -> at this point in time it will use the strategy from the CR `EvictionStrategyNone`
- I change the eviction strategy in KubeVirt CR to `EvictionStrategyLiveMigrate`
- the VM will use the strategy from the CR, i.e. `EvictionStrategyLiveMigrate`

With the suggested defaulting, the behavior changes:
- initially the eviction strategy in KubeVirt CR is `EvictionStrategyNone`
- I create a VM with an empty eviction strategy -> `vmi.spec.evictionStrategy` is defaulted to `EvictionStrategyNone`
- I change the eviction strategy in KubeVirt CR to `EvictionStrategyLiveMigrate`
- the VM will still use `EvictionStrategyNone` since it is defaulted in the VMI spec.

Do I miss smth?

Comment (acardace) [reply pkg/virt-controller/watch/vmi.go:608]: > > So I dont think it will be a behavior change
> 
> Still, I am not sure. Looking at the code:
> 
> https://github.com/kubevirt/kubevirt/blob/0a91f46054c6429cd4311e4ecc255bdefe442b2d/pkg/util/migrations/migrations.go#L65-L71
> 
> So, the current flow:
> 
> * let's say, initially the eviction strategy in KubeVirt CR is `EvictionStrategyNone`
> * I create a VM with an empty eviction strategy -> at this point in time it will use the strategy from the CR `EvictionStrategyNone`
> * I change the eviction strategy in KubeVirt CR to `EvictionStrategyLiveMigrate`
> * the VM will use the strategy from the CR, i.e. `EvictionStrategyLiveMigrate`

It is not really like this because of one detail, let's say you specify `evictionStrategy` to `LiveMigrate` at cluster-level, you start the VM and then you later change the KV CR field to `None`, if you try to evict the VM the virt-launcher pod won't still be evicted as the corresponding PDB won't be deleted by virt-controller in reaction to the `evictionStrategy` change (as it should be), hence practically the current behaviour retains the initial `evictionStrategy`. To actually change it you have to restart the VM like most things in the VM object.

That said you actually uncovered a bug that would be fixed by using the `spec` instead of the `status` for storing the initial `evictionStrategy` as right now in virt-controller if we face the situation I just described the `evictionStrategy` for the VM will be evaluated to `None` after the CR change even if it actually behaves as `LiveMigrate`.

TLDR: using the spec is definitely the correct choice here both in terms of observability and behaviour.  

> 
> With the suggested defaulting, the behavior changes:
> 
> * initially the eviction strategy in KubeVirt CR is `EvictionStrategyNone`
> * I create a VM with an empty eviction strategy -> `vmi.spec.evictionStrategy` is defaulted to `EvictionStrategyNone`
> * I change the eviction strategy in KubeVirt CR to `EvictionStrategyLiveMigrate`
> * the VM will still use `EvictionStrategyNone` since it has been defaulted in the VMI spec.

This is indeed the behaviour we want.

Comment (vasiliy-ul) [reply pkg/virt-controller/watch/vmi.go:608]: Hm, I see. Probably I was mostly looking at it from the API perspective without taking into account other aspects. Thank you for clarifying it!

**Comment by acardace** [comment]
@vasiliy-ul @alaypatel07 PR updated PTAL and thank you for all the comments.

File: pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go [hunk 3]
```diff
@@ -1153,4 +1153,46 @@ var _ = Describe("VirtualMachineInstance Mutator", func() {
 				},
 			}),
 	)
+
+	DescribeTable("evictionStrategy should match the", func(f func(*v1.VirtualMachineInstanceSpec) v1.EvictionStrategy) {
+		expected := f(&vmi.Spec)
+		_, vmiSpec, _ := getMetaSpecStatusFromAdmit(rt.GOARCH)
+		Expect(*vmiSpec.EvictionStrategy).To(Equal(expected))
```
Comment (vasiliy-ul) [inline pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go:1160]: nit:

```suggestion
                Expect(vmiSpec.EvictionStrategy).ToNot(BeNil())
		Expect(*vmiSpec.EvictionStrategy).To(Equal(expected))
```

File: pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go [hunk 4]
```diff
@@ -1153,4 +1153,46 @@ var _ = Describe("VirtualMachineInstance Mutator", func() {
...
+			return vmiStrategy
+		}),
+		Entry("default one if nothing is set", func(s *v1.VirtualMachineInstanceSpec) v1.EvictionStrategy {
+			s.EvictionStrategy = nil
+
+			kvCR := testutils.GetFakeKubeVirtClusterConfig(kvInformer)
+			kvCR.Spec.Configuration.EvictionStrategy = nil
+			testutils.UpdateFakeKubeVirtClusterConfig(kvInformer, kvCR)
+
+			return *mutator.ClusterConfig.GetDefaultClusterConfig().EvictionStrategy
```
Comment (vasiliy-ul) [inline pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go:1195]: nit:
```suggestion
                        defaultEvictionStrategy := mutator.ClusterConfig.GetDefaultClusterConfig().EvictionStrategy
                        Expect(defaultEvictionStrategy).ToNot(BeNil())
			return *defaultEvictionStrategy
```

**Review by vasiliy-ul — COMMENTED** [review]
Looks good. Just added a couple of nits inline.

File: pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go (hunk 3 above)
Comment (acardace) [reply pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go:1161]: done

File: pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go (hunk 4 above)
Comment (acardace) [reply pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go]: done

File: pkg/virt-controller/watch/vmi.go (hunk 2 above)
Comment (alaypatel07) [reply pkg/virt-controller/watch/vmi.go]: +1, this is the exact behavior we wanted with vmi.spec.architecture and its the right way to default optional fields that dont change over the lifetime of the vmi IMO

**Review by alaypatel07 — APPROVED** [review]
LGTM from api perspective
